File: CPP/BFS/Maze-Traversal.cpp
#include<iostream>
#include<queue>
#include<cstring>

using namespace std;

struct Node{
    int x,y,dist;
};

int row[4] = {-1,0,0,1};
int col[4] = {0,1,-1,0};

bool isValid(int s1,int s2, int i, int j, char maze[][5]){
    return s1 > 0 && s2 > 0 && maze[s1][s2] != '#' && maze[i][j] != NULL && maze[i][j] != '#';
}

void BFS(char maze[][5],int s1, int s2, int d1, int d2){
    bool visited[5][5];
    int distance = 0,a,b;
    memset(visited, false, sizeof(visited));
    queue<Node> q;
    Node node;
    node.x = s1;
    node.y = s2;
    node.dist = 0;
    q.push(node);
        while(!q.empty()){
            Node temp = q.front();
            q.pop();
            a = temp.x;
            b = temp.y;
            if(temp.x == d1 && temp.y ==d2){
                //return 
            }
            for(int i = 0; i < 4; i++){
                if(isValid(a, b, a + row[i], b + col[i], maze)){
                    temp.dist = temp.dist + 1;
                    temp.x = a + row[i];
                    temp.y = b + col[i];
                    q.push(temp);
                }
            }

        }

}

void displayMaze(char maze[][5]){
    for(int i = 0; i<5; i++){
        for(int j=0; j<5; j++){
            cout << maze[i][j] <<" ";
        }
        cout << "\n";
    }
}

int main(){

    char maze[][5] = {
                            {'.','.','.','.','.'},
                            {'.','.','#','.','.'},
                            {'#','.','.','#','.'},
                            {'.','#','.','.','.'},
                            {'.','.','#','.','.'}    
                      };
    displayMaze(maze);
    BFS(maze,1,3,3,3);
    return 0;
}

File: Java/BFS/MazeTraversal.java
import java.io.*;
import java.util.Arrays;
import java.util.Queue;
import java.util.LinkedList;


class Node{
    int x,y,dist;
}



public class MazeTraversal{
                    // D  U R L
    static int[] x1 = {1,-1,0, 0};
    static int[] y1 = {0, 0,1,-1};

    void displayMaze(boolean[][] maze){
        for(int i = 0; i < 4; i++){
            for(int j = 0; j < 5; j++){
                System.out.print(maze[i][j] + " ");
            }
            System.out.println("\n");
        }
    }


    boolean isValid(int p, int q, int dist, boolean[][] booleanMaze, int[][] maze){
        
        if(p >= 0 && q >= 0 && dist >=0 && p <= 3 && q <= 4){
            if(maze[p][q] != 0 && booleanMaze[p][q]){
                return true;
            }else{
                return false;
            }
        }else{
            return false;
        }
    }

    Node shortestPath(int[][] maze, int s1, int s2, int d1, int d2){
        boolean[][] booleanMaze  = new boolean[4][5];
        for(boolean[] row : booleanMaze){
            Arrays.fill(row, true);
        }

        //displayMaze(booleanMaze);
            Node node = new Node();
            int p,q,dist;
            node.x = s1;
            node.y = s2;
            node.dist = 0;
            Queue<Node> que = new LinkedList<>();
            que.add(node);
            booleanMaze[3][4] = false;
                
                while(!que.isEmpty()){

                    Node point = que.peek();
                    Node dest = new Node();
                    for(int i=0 ; i < 4 ; i++){
                        p = point.x + x1[i];
                        q = point.y + y1[i];
                        dist = point.dist;
                        System.out.println("####");
                        //System.out.print(booleanMaze[point][q]);
                        System.out.println("***  (" + p + "," + q +")"+ " dist:" + dist);
                        if(isValid(p, q, dist, booleanMaze, maze)){
                            //System.out.println("$$");
                            
                            dest.x = p;
                            dest.y = q;
                            dest.dist = point.dist + 1;
                            que.add(dest);
                            
                            System.out.println(" $$ (" + dest.x + "," + dest.y +")"+ " dist:" + dest.dist);
                            
                        }
                        if(p == d1 && q==d2){
                            return dest;
                        }
                        if(p >= 0 && q >=0 && p <= 3 && q <= 4){
                            booleanMaze[p][q] = false;
                        }

                    }
                    que.remove();
                }

        return null;
    }

    public static void main(String args[]){
        int[][] maze = {{1,1,0,1,1},
                        {0,1,0,1,1},
                        {0,1,1,1,0},
                        {1,0,1,1,1}};

        MazeTraversal m1 = new MazeTraversal();
        Node destinationNode = m1.shortestPath(maze,3,4,1,4);
        if(destinationNode !=  null){
            System.out.println("Shortest Path : " + destinationNode.dist);
        }else{
            System.out.println("No Shortest path found");
        }
    }
}